ssl/ssl_lib.c

```diff
@@ -6901,3 +6909,173 @@ int SSL_CTX_set0_tmp_dh_pkey(SSL_CTX *ctx, EVP_PKEY *dhpkey)
     ctx->cert->dh_tmp = dhpkey;
     return 1;
 }
+
+#ifndef OPENSSL_NO_COMP
+
+static int ssl_set_cert_comp_pref(uint16_t *array, size_t array_size, uint16_t *algs, int num_algs)
+{
+    int i, j = 0;
+
+    /* This will 0-terminate the array */
+    memset(array, 0, array_size);
+    /* include only those algorithms we support, ignore the rest if the list fills */
+    for (i = 0; i < num_algs && j < TLSEXT_comp_cert_limit; i++) {
+#if defined(BROTLI)
+        if (algs[i] == TLSEXT_comp_cert_brotli)
+            array[j++] = algs[i];
+#endif
+#if defined(ZLIB)
+        if (algs[i] == TLSEXT_comp_cert_zlib)
+            array[j++] = algs[i];
+#endif
+#if defined(ZSTD)
+        if (algs[i] == TLSEXT_comp_cert_zstd)
+            array[j++] = algs[i];
+#endif
+    }
```

Comment:
No, because it's perfectly acceptable to have an input of:
```
[ brotli, zlib, zstd, some-fangled-new-compression-not-yet-supported ]
```
The `array` will fill up, but it's legal, and the last value will be ignored.
This could only happen if the array has duplicate entries, which this doesn't check for, and it should because I don't see how the following would be considered legal:
```
[ brotli, zsts, brotli, zlib ]

Reply:
Illegal inputs are not detected.